runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -1493,6 +1500,201 @@ J9::CodeGenerator::lowerTreeIfNeeded(
    }
 
 
+/*
+ * If value types are enabled, and the value that is being assigned to the array
+ * element might be a null reference, lower the ArrayStoreCHK by splitting the
+ * block before the ArrayStoreCHK, and inserting a NULLCHK guarded by a check
+ * of whether the array's component type is a value type.
+ */
+void
+J9::CodeGenerator::lowerArrayStoreCHK(TR::Node *node, TR::TreeTop *tt)
+   {
+   TR::Node *destChild;
+   TR::Node *sourceChild;
+
+   // Pattern match the ArrayStoreCHK operands to get the source of the assignment
+   // (sourceChild) and the array to which an element will have a value assigned (destChild)
+   findArrayStoreCHKOperands(node, destChild, sourceChild);
+
+   // Only need to lower if it is possible that the value is a null reference
+   if (!sourceChild->isNonNull())
+      {
+      TR::CFG * cfg = self()->comp()->getFlowGraph();
+      cfg->invalidateStructure();
+
+      TR::Block *prevBlock = tt->getEnclosingBlock();
+
+      performTransformation(self()->comp(), "%sTransforming ArrayStoreCHK n%dn [%p] by splitting block block_%d, and inserting a NULLCHK guarded with a check of whether the component type of the array is a value type\n", OPT_DETAILS, node->getGlobalIndex(), node, prevBlock->getNumber());
+
+      // Anchor the node that contains the destination array and the node
+      // containing the source of the array element assignment to ensure
+      // they are available in the ificmpeq and NULLCHK
+      TR::TreeTop *anchoredArrayTT = TR::TreeTop::create(self()->comp(), tt->getPrevTreeTop(), TR::Node::create(TR::treetop, 1, destChild));
+      TR::TreeTop *anchoredSourceTT = TR::TreeTop::create(self()->comp(), anchoredArrayTT, TR::Node::create(TR::treetop, 1, sourceChild));
+
+      // Transform
+      //   +--------------------------------+
+      //   | ttprev                         |
+      //   | ArrayStoreCHK                  |
+      //   |   awrtbari                     |
+      //   |     aladd                      |
+      //   |       <array-reference>        |
+      //   |       index-offset-calculation |
+      //   |     <value-reference>          |
+      //   +--------------------------------+
+      //
+      // into
+      //   +--------------------------------+
+      //   | treetop                        |
+      //   |   <value-reference>            |
+      //   | treetop                        |
+      //   |   <array-reference>            |
```

Comment:
The array reference needs to be evaluated before the value reference since the array-reference is deeper.

Reply:
Thanks for catching that!

Reply:
Thanks, Andrew.  My comment didn't match the code - fixed in commit https://github.com/eclipse/openj9/commit/58ab83bcf1cf130105931a3c27f9fa299cc3acd4